### DISCUSSION TITLE
How can i use useObservable to create a subject

### DISCUSSION THREAD

**Discussion by ShravanSunder** [crimx/observable-hooks#49] [Q&A] [answered]
Hi! I'm currently learning RxJs and using it with react.  

` const flag$ = useObservable(pluckFirst, [props.flag])`

I was wondering how i can use the above to create a subject?  I wanted to do it while staying in react.  Do i have to create a separate subject resource thats outside of react?

thanks for the wonderful library

**Comment by crimx** [comment, marked as answer]
Yes you can! "[Do whatever you want with Observables](https://observable-hooks.js.org/)" is one of the main features of observable-hooks.

If you want to create Subject from dependencies, you can do this:

```ts
const flag$ = useObservable(
  inputs$ => inputs$.pipe(
    map(([flag]) => flag),
    share()
  ),
  [props.flag]
)
```

Or if you just want to create a Subject out of thin air, you can do this:

```ts
const subject$ = useObservable(() => new Subject())
```

**Reply by ShravanSunder** [reply]
wow great!  If you don't mind I'd like to contribute to the docs as I get familiar with this library.

The second option was what I was looking for!

**Reply by crimx** [reply]
That would be great! PRs are welcome! 🤩

Also you might wanna take a look at the [`useObservable`](https://observable-hooks.js.org/api/#useobservable) API docs. There are explanations and examples on why and how to use it.

**Reply by ShravanSunder** [reply]
@crimx as a follow-up question, is there a reason the API doesn't have something like useSubject to pass back a subject?

Do you see any problems with the below code? if I wanted to make a hook to access a subject?  I am using the `next ` method in the subject to pass data into the subject from child components

```js

/**
 * returns
 * 1. an observable of a value
 * 2. the current state
 */
export type SubjectWithState<T> = { subject$: BehaviorSubject<T> | undefined; state: T };


/**
 * Returns an observable from initial state and current state
 * @param value
 */

/**
 * Returns an observable from initial state and current state
 * @param value
 */
const useCreateSubject = <T1,>(value: T1): SubjectWithState<T1> => {
   const observable$ = useObservable(pluckFirst, [value]);
   const state = useObservableState(observable$, value);

   return { subject$: observable$ as BehaviorSubject<T1>, state };
};


```

**Reply by crimx** [reply]
Were you trying to achieve something like the  [`useObservableCallback`](https://observable-hooks.js.org/api/#useobservablecallback)?

**Reply by ShravanSunder** [reply]
I did think about using ` useObservableCallback` and it seemed like something i could use.  

However i wanted to reduce the number of properties i would use to access the state, observable and to send data to the observable as well.  So seemed like subject with state was a good option. 

I'm still experimenting with the above to see if it fits my need for local state with streams and actions.

**Reply by ShravanSunder** [reply]
in the end i used used `useObservableCallback?` as you mentioned.  Thanks for the help!
